### DISCUSSION TITLE
消费者负载均衡有些疑惑

### DISCUSSION THREAD

**Discussion by xhrg** [apache/rocketmq#8588] [General]
rocketmq的消费者负载均衡，是通过多个节点维护同一个算法，让各自计算得到不同的queue然后去拉消息。

请问是否存在一种情况，2个节点因为时间先后，算到同一个queue，那这种情况下，会怎么样，rocketmq是如何解决该问题的。

**Comment by make-suaron** [comment]
可以去了解一下 rocketmq 中的 rebalance 机制

**Reply by make-suaron** [reply]
https://developer.aliyun.com/article/1337958

**Comment by cserwen** [comment]
你说的这种情况确实存在 @xhrg 
- 在4.x 版本的 Rebalance 算法中，得到的效果是最终一致，rebalance 期间会出现短暂的堆积或者消费重复，但是由于 Broker 会同时通知所有 client 进行 balance，所以这个耗时会非常短，影响不大
- 到了5.x版本，我们支持了 POP 消费，拉取消息时服务端会加锁，消费进度由服务端维护，不会出现一个queue同时被多个节点消费的情况
